express-notes-api/server.ts

```diff
@@ -0,0 +1,131 @@
+import { readFile, writeFile } from 'node:fs/promises';
+import express from 'express';
+
+const app = express();
+
+app.use(express.json());
+
+app.get('/api/notes', async (req, res) => {
+  try {
+    const notesArray = await read();
+    res.status(200).json(notesArray);
+  } catch (error) {
+    res.status(404).json({ error: `${error.message}` });
+  }
+});
+
+app.get('/api/notes/:id', async (req, res) => {
+  try {
+    const { notes } = await readData();
+    const requestId = Number(req.params.id);
+    if (requestId < 0) {
+      res.status(400).json({ error: 'ID needs to be a positive number' });
+      return;
+    }
+    if (!notes[requestId]) {
+      res.status(404).json({ error: `ID ${requestId} not in notes` });
+      return;
+    }
+    res.status(200).json(notes[requestId]);
+  } catch (error) {
+    res.status(404).json({ error: `${error.message}` });
+  }
+});
+
+app.post('/api/notes', async (req, res) => {
+  try {
+    let { nextId, notes } = await readData();
+    const { content } = req.body;
+    if (!content) {
+      res.status(400).send({ error: 'content is a required field' });
+      return;
+    }
+    try {
+      const newNote = { id: nextId, content };
+      notes[nextId] = newNote;
+      nextId++;
+      await writeData({ nextId, notes });
+      res.status(201).json(newNote);
+    } catch (error) {
+      console.error(error);
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+    }
+  } catch (error) {
+    res.status(404).json({ error: `${error.message}` });
```

Comment:
Why do you have nested try/catch blocks? It works, but I'm not understanding why you did it. And even more strange, you send a 404 from this one, which is where an error in `readData` would end up.

Reply:
I had this weird and incorrect perception that I had to send back a 400's status code if it fails to read the file and 500's code if it fails to write the file, so I put readData and writeData in different try/catch blocks so they can return different status code. Since they're both server side code I think they would both just get the 500 status code so I can put them all in one try/catch block.